### PR TITLE
feat(fidelity-harness): add per-field null fidelity check

### LINKED ISSUE #4917: Null handling ensure the serdes handle null fields
Expand AllMessagesFidelityCheckTest or add a new test covering every field in in every message to ensure we can handle null's for every field.

### PR DESCRIPTION
**Pull request by SamBarker**
### Type of change

- Enhancement / new feature

### Description

Adds fidelity coverage for null handling: a new `NullableFieldFidelityTest` with two parameterised tests in `kroxylicious-wire-fidelity-tests`. For every (message, direction, version, nullable field) combination the Kafka protocol schema defines, it nulls out exactly that one field and checks the message round-trips correctly in both directions (Kroxylicious reads Kafka-serialised bytes, Kafka reads Kroxylicious-serialised bytes).

To drive this, `kroxylicious-fidelity-harness` gains:
- `NullableFieldEnumerator` — walks a message's Kafka runtime `Schema` at a given version and returns every nullable `BoundField`, recursing into nested structs and arrays-of-structs (not just top-level fields).
- `NullFieldStrategy` — a `FieldPopulationStrategy` that nulls one target field (matched by `BoundField` reference identity, since the same static per-version schema field instances are produced by both the enumerator and the populator) and delegates every other field to a wrapped strategy (`ValidScalarStrategy`).

`AllMessagesFidelityCheckTest`'s message/version enumeration logic was extracted into a shared `AllMessages` helper so the new test can reuse it without duplicating ~100 lines of reflection/`ApiKeys` plumbing.

Across the full protocol surface (604 message/direction/version combinations), 489 (message, version, field) tuples are nullable at that version — most combinations have no nullable fields at all, since nullability was introduced incrementally per KIP and most fields are non-nullable primitives. That yields 978 test invocations total (489 × the 2 read directions).

**In scope:** fields whose *type* carries an explicit on-wire null representation (`Type.isNullable()`), i.e. `NULLABLE_STRING`/`NULLABLE_BYTES`/`NULLABLE_RECORDS` and their compact variants, including such fields nested inside structs and arrays-of-structs.

**Out of scope:** flexible-version *tagged fields* that are omitted from the wire and take a default value on read. That's a distinct mechanism (presence/absence, not null-encoding) and is tracked separately in #4927.

### Additional Context

Closes #4917 (sub-issue of #4916, "Build out the fidelity harness").

### Checklist

- [x] New tests written (where applicable).
- [ ] N/A — no user-facing/documentation change (test harness only).
- [x] Existing unit/integration/system tests passing.
- [ ] Sonarcloud warnings addressed (`@SuppressWarnings("ReferenceEquality")` added with justifying comment on the reference-identity match in `NullFieldStrategy`).
- [x] PR references related GitHub issue(s) so they are closed on merging.
- [x] If AI tools assisted with code changes, commit messages include `Assisted-by:` trailer.
- [ ] Changelog entry — none added; this only changes internal test-fixture code (test artefacts, not published API consumed by end users).

### CODE REVIEW

File: kroxylicious-wire-fidelity-tests/src/test/java/io/kroxylicious/fidelity/kafka/NullableFieldFidelityTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        PopulationResult populated = populator.populate(kafkaMessage, version);
+
+        // When
+        ReadResult<?> result = FidelityCheck.kroxyliciousReads(
+                kafkaMessage,
+                (io.kroxylicious.kafka.common.protocol.Message) kroxyliciousMessage,
+                version);
+
+        // Then
+        assertThat(populated).isInstanceOf(PopulationResult.Populated.class);
```
Comment (robobario) [inline kroxylicious-wire-fidelity-tests/src/test/java/io/kroxylicious/fidelity/kafka/NullableFieldFidelityTest.java:53]: debugging this I saw an odd behaviour. looking at RequestData null cluster_id I see:

1. kroxyliciousMessage and kafkaMessage both have initially have null clusterId 
2. populator.populate is called, after which point `kafkaMessage` has a randomized clusterId `YMKsHHVG`.
3. kroxyliciousReads is called on the input kafkaMessage which has a non-null clusterId
4. result.message has a non-null clusterId

test passes because kafkaMessage matches result message.

Which feels like something has gone wrong here: claude pointed out:

```
SchemaDrivenMessagePopulator.expandFields (line 93) does new BoundField(entry.getValue(), null, entry.getKey()) for each tagged field, every call. NullableFieldEnumerator.nullableFields captures target
  BoundField from one expandFields call. populate() later calls expandFields again → new instance. NullFieldStrategy.resolve uses field == target — always false for tagged fields.

  Confirmed reachable: FetchRequestData/FetchSnapshotRequestData have nullable tagged cluster_id (COMPACT_NULLABLE_STRING in TaggedFieldsSection). Test case "null cluster_id" silently populates valid value
  instead, still passes. Defeats whole point of per-field null fidelity test for every tagged nullable field in protocol.
```

and fixed with:

```
diff --git a/kroxylicious-fidelity-harness/src/main/java/io/kroxylicious/fidelity/populate/NullFieldStrategy.java b/kroxylicious-fidelity-harness/src/main/java/io/kroxylicious/fidelity/populate/NullFieldStrategy.java
index 313dad2ea..666e86949 100644
--- a/kroxylicious-fidelity-harness/src/main/java/io/kroxylicious/fidelity/populate/NullFieldStrategy.java
+++ b/kroxylicious-fidelity-harness/src/main/java/io/kroxylicious/fidelity/populate/NullFieldStrategy.java
@@ -12,10 +12,13 @@ import org.apache.kafka.common.protocol.types.BoundField;
  * strategy.
  * <p>
  * Matching by reference rather than by name sidesteps the fact that {@link BoundField} carries no
- * ancestry: {@link NullableFieldEnumerator} hands back the actual {@link BoundField} instances from
- * Kafka's static, per-version {@code Schema}, and the same instances are produced again when
- * {@link SchemaDrivenMessagePopulator} walks that schema to populate a message, so identity is a
- * reliable way to target the exact field that was enumerated.
+ * ancestry. Matching is done on {@link BoundField#def} rather than the {@link BoundField} itself: for a
+ * tagged field, {@link SchemaDrivenMessagePopulator#expandFields} allocates a fresh {@link BoundField}
+ * wrapper on every call, so two enumerations of the same tagged field never share a {@link BoundField}
+ * instance. The underlying {@link org.apache.kafka.common.protocol.types.Field} each wrapper carries,
+ * however, comes from {@link org.apache.kafka.common.protocol.types.TaggedFields}'s own field map, which
+ * is built once and never rebuilt - so it, unlike the wrapper, is a reliable identity to target the exact
+ * field that was enumerated, whether tagged or not.
  */
 public final class NullFieldStrategy implements FieldPopulationStrategy {
 
@@ -34,8 +37,8 @@ public final class NullFieldStrategy implements FieldPopulationStrategy {
     }
 
     @Override
-    @SuppressWarnings("ReferenceEquality") // BoundField declares no equals()/hashCode(); identity is the only, and intended, notion of equality here.
+    @SuppressWarnings("ReferenceEquality") // Field declares no equals()/hashCode(); identity is the only, and intended, notion of equality here.
     public FieldDecision resolve(BoundField field) {
-        return field == target ? new FieldDecision.Value(null) : delegate.resolve(field);
+        return field.def == target.def ? new FieldDecision.Value(null) : delegate.resolve(field);
     }
 }

```

**Comment by robobario** [comment]
issue created for unrelated test failure https://github.com/kroxylicious/kroxylicious/issues/4929

File: kroxylicious-wire-fidelity-tests/src/test/java/io/kroxylicious/fidelity/kafka/NullableFieldFidelityTest.java (hunk 1 above)
Comment (SamBarker) [reply kroxylicious-wire-fidelity-tests/src/test/java/io/kroxylicious/fidelity/kafka/NullableFieldFidelityTest.java:53]: Thanks for catching that one Rob.

File: kroxylicious-fidelity-harness/src/main/java/io/kroxylicious/fidelity/populate/SchemaDrivenMessagePopulator.java [hunk 2]
```diff
@@ -379,16 +376,11 @@ private static Object convertToCollectionShape(List<?> elements, Method setter,
...
-        try {
-            @SuppressWarnings("unchecked")
-            Collection<Object> collection = (Collection<Object>) collectionType.getDeclaredConstructor().newInstance();
-            collection.addAll(elements);
-            return collection;
-        }
-        catch (ReflectiveOperationException e) {
-            throw new IllegalStateException("Could not instantiate collection " + collectionType, e);
-        }
+        Collection<Object> collection = (Collection<Object>) AllMessages.newInstance(collectionType, Collection.class);
```
Comment (robobario) [inline kroxylicious-fidelity-harness/src/main/java/io/kroxylicious/fidelity/populate/SchemaDrivenMessagePopulator.java:381]: nit: feels a little odd that AllMessages is about messages but we're using it as a reflection utility, not sure it's worth pulling out another reflection util class though.

File: kroxylicious-wire-fidelity-tests/src/test/java/io/kroxylicious/fidelity/kafka/NullableFieldFidelityTest.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        return nullableFields.stream()
+                .map(field -> Arguments.argumentSet(versionedMessage.label() + " - null " + field.def.name,
+                        versionedMessage.version(),
+                        freshInstance(versionedMessage.kroxyliciousMessage()),
+                        freshInstance(versionedMessage.kafkaMessage()),
+                        field));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> T freshInstance(T template) {
```
Comment (robobario) [inline kroxylicious-wire-fidelity-tests/src/test/java/io/kroxylicious/fidelity/kafka/NullableFieldFidelityTest.java]: nit: this looks similar to the AllMessages#newInstance, we could deduplicate

**Review by robobario — APPROVED** [review]
LGTM with nits

File: kroxylicious-fidelity-harness/src/main/java/io/kroxylicious/fidelity/populate/SchemaDrivenMessagePopulator.java (hunk 2 above)
Comment (SamBarker) [reply kroxylicious-fidelity-harness/src/main/java/io/kroxylicious/fidelity/populate/SchemaDrivenMessagePopulator.java:381]: Calude tried to add a new reflection class on exactly the grounds you mention and I told it put it here. If I squint its in the right place as this AllMessages should be responsible for creating message instances... but its not a strong argument...
